src/apps/tools.clj

```diff
@@ -58,19 +60,29 @@
          (add-query-offset offset)
          (add-hidden-tool-types-clause include-hidden)))))
 
+(defn format-tool-listing
+  [perms public-tool-ids {:keys [id] :as tool}]
+  (-> tool
+      (assoc :is_public  (contains? public-tool-ids id)
+             :permission (or (perms id) ""))
+      remove-nil-vals))
+
 (defn search-tools
   "Obtains a listing of tools for the tool search service."
-  [params]
-  {:tools
-     (map remove-nil-vals
-       (select (tool-listing-base-query params)))})
+  [{:keys [user] :as params}]
+  (let [perms           (permissions/load-tool-permissions user)
+        public-tool-ids (permissions/get-public-tool-ids)]
+    {:tools
+     (map (partial format-tool-listing perms public-tool-ids)
+       (select (tool-listing-base-query params)))}))
 
 (defn get-tool
   "Obtains a tool by ID."
-  [tool-id]
+  [user tool-id]
   (let [tool           (->> (first (select (tool-listing-base-query) (where {:tools.id tool-id})))
                             (assert-not-nil [:tool-id tool-id])
-                            remove-nil-vals)
+                            (format-tool-listing (permissions/load-tool-permissions user)
+                                                 (permissions/get-public-tool-ids)))
```

Comment:
I don't see how this function prevents users from viewing tools that they're not permitted to view either.